### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by Warhammer4000**
Known issue
-Event subscription bug : Trying to play new dialogue before one ends ends up creating a bug that breaks the text display

### CODE REVIEW

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/Data/CharacterTableEditor.cs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+using Studio23.SS2.DialogueSystem.Data;
+
+namespace com.studio23.ss2.dialoguesystem.editor
+{
+    public class CharacterTableEditorWindow : EditorWindow
+    {
+        private CharacterTable characterTable;
+        private string characterID = "";
+        private string characterName = "";
+        private string csvFilePath = "";
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/Data/CharacterTableEditor.cs]: Did not follow the variable naming convention.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueManager.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+
+
+
+
+
+
+
+
+    }
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueManager.cs]: Unnecessary whitespace.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueManager.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
+
+using Studio23.SS2.DialogueSystem.Data;
+using Studio23.SS2.DialogueSystem.UI;
+using UnityEngine;
+
+namespace Studio23.SS2.DialogueSystem.Core
+{
+    public class DialogueManager : MonoBehaviour
+    {
+
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueManager.cs:10]: Unnecessary space.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+using UnityEngine.UI;
+
+
+namespace Studio23.SS2.DialogueSystem.UI
+{
+
+    public class DialogueUI : MonoBehaviour
+    {
+        [Header("UI Components")]
+        public TextMeshProUGUI dialogueText;
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: Did not follow naming convention rules.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        [SerializeField]private DialogueState dialogueState;
+
+        [Header("Configuration")]
+        [SerializeField]private DialogueUIConfig config;
+
+
+
+      
+
+        private CancellationTokenSource CancelCurrentDialogueLineToken;
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: Did not follow naming convention rules.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+
+        void Start()
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: Must explicitly mention access modifiers.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+
+        public void RegisterGraph(DialogueGraph newGraph)
+        {
+            _currentGraph = newGraph;
+            _currentGraph.ClearEvents();
+            _currentGraph.OnDialogueStart += () => PlayDialogue();
+            _currentGraph.OnDialogueComplete += () => EndDialogue();
+            _currentGraph.OnDialogueNext += async () => await ShowDialogueText(_currentGraph.CurrentNode.DialogueText);
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: Lacks readability considering everyone is still not familiar with lambda expression short-hands mixed with async function calls.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Configuration/DialogueUIConfig.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    [CreateAssetMenu(fileName = "Studio-23", menuName = "Studio-23/Dialogue System/New UI Config")]
+    public class DialogueUIConfig : ScriptableObject
+    {
+        [Header("Text Display Settings")]
+        public float letterDelay = 0.05f;
+        public float nextSentenceDelayMultiplier = 1.0f;
+
+        [Header("Subtitle Configuration")]
+        public Color subtitleColor = Color.black;
+        public int subtitleFontSize = 24;
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Configuration/DialogueUIConfig.cs]: Did not follow naming convention rules.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+{
+    //todo remove this
+    [CreateAssetMenu(menuName = "Studio-23/Dialogue System/New Character Table")]
+    public class CharacterTable : ScriptableObject
+    {
+        [Serializable]
+        public class CharacterInfo
+        {
+            public string characterID;
+            public string characterName;
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs]: Did not follow naming convention rules.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+            public List<CharacterReactionImage> Reactions;
+        }
+
+        [Serializable]
+        public class CharacterReactionImage
+        {
+            public string Reaction;
+            public Sprite Image;
+        }
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs:25]: Let's keep every class under its separate script unless it is only used within that script and nowhere else.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+        public List<CharacterInfo> characterList = new List<CharacterInfo>();
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs:48]: Did not follow naming convention rules.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs [hunk 12]
```diff
@@ -0,0 +1,35 @@
+﻿
+
+
+using UnityEngine;
+using XNode;
+
+
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs]: Extra whitespaces were not removed.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs:1]: File contains many vertical whitespaces which are unnecessary.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueNode.cs
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueNode.cs:1]: Contains unnecessary and extra vertical spaces which needs to be removed.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/Data/DialogueBaseNodeEditor.cs [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Studio23.SS2.DialogueSystem.Data
+{
+    [CustomNodeEditor(typeof(DialogueBase))]
+    public class DialogueBaseNodeEditor : NodeEditor
+    {
+
+        public Color TextColor= Color.white;
+        private DialogueBase dialogueBase;
+        private static GUIStyle editorLabelStyle;
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/Data/DialogueBaseNodeEditor.cs:13]: Did not follow naming convention rules.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+            dialogueText.text = string.Empty;
+            dialogueState = DialogueState.Started;
+            CancelCurrentDialogueLineToken = new CancellationTokenSource();
+
+            ShowUI(true);
+            for (int i = 0; i < text.Length; i++)
+            {
+                await UniTask.Delay((int)(config.letterDelay * 1000), cancellationToken: CancelCurrentDialogueLineToken.Token).SuppressCancellationThrow();
+                if (dialogueState == DialogueState.Skipped) break;
+                dialogueText.text += text[i];
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: Can be done without string concat via textmeshpro max visible char count. Would prevent UI resizing issues as well.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        {
+
+            dialogueText.text = string.Empty;
+            dialogueState = DialogueState.Started;
+            CancelCurrentDialogueLineToken = new CancellationTokenSource();
+
+            ShowUI(true);
+            for (int i = 0; i < text.Length; i++)
+            {
+                await UniTask.Delay((int)(config.letterDelay * 1000), cancellationToken: CancelCurrentDialogueLineToken.Token).SuppressCancellationThrow();
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: `config.letterDelay * 1000`, using arbitrary values as such creates a lack of readability in the code. Use properly named variables to ensure that other developers are able to understand the purpose. Alternatively, there's `UniTask.Delay(Timespan.FromSeconds)` and other similar methods to use in these cases.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+                if (dialogueState == DialogueState.Skipped) break;
+                dialogueText.text += text[i];
+
+            }
+
+            await UniTask.Delay((int)(text.Length * 100 * config.nextSentenceDelayMultiplier), cancellationToken: CancelCurrentDialogueLineToken.Token).SuppressCancellationThrow();
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: using arbitrary values as such creates a lack of readability in the code. Use properly named variables to ensure that other developers are able to understand the purpose. Alternatively, there's `UniTask.Delay(Timespan.FromSeconds)` and other similar methods to use in these cases.

**Review by N-Evan — APPROVED** [review]
The overall PR contains formatting and conventional issues. No major functional issues were noticed in the review so far hence approving it. 

I expect to see the commented changes in the upcoming PR and no further incorrect naming style and unnecessary whitespace in scripts. 

Edit: Additionally, please initiate the repository with the initial project files and necessary dependencies to avoid unnecessarily large PRs like this one. It would also be helpful if feature branches were named accordingly instead of a generic name like dev-x or feature. 

The PR title lacks understandability as well. With something generic as `Initial Commit` as PR title, there's no way for a reviewer or another developer to understand what was done without having to go through the code line-by-line. A bit more descriptive and functional titles are expected in the future.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+using System.Threading;
+using TMPro;
+using UnityEngine;
+using UnityEngine.UI;
+
+
+namespace Studio23.SS2.DialogueSystem.UI
+{
+
+    public class DialogueUI : MonoBehaviour
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: We will need to change the UI so having a base class with overrideable PlayDialogue, EndDialogue, ShowDialogueText would be useful

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueManager.cs [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+        public void ChangeDialogueGraph(DialogueGraph newGraph)
+        {
+            _currentGraph = newGraph;
+        }
+
+        public void PlayDialogue()
+        {
+            DialogueUI.RegisterGraph(_currentGraph);
+            _currentGraph.StartDialogue();
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueManager.cs:39]: Dialoguemanger should have events that fire when dialogue starts and ends so that rest of the systems can hook in without needing to know which graph initiated the dialogue.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            public string characterID;
+            public string characterName;
+            public List<CharacterReactionImage> Reactions;
+        }
+
+        [Serializable]
+        public class CharacterReactionImage
+        {
+            public string Reaction;
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs:23]: Assuming we want to pick reactions per dialogue node, enums for reactions would make it less error prone to select.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+        private void ApplyConfiguration()
+        {
+            dialogueText.fontSize = config.subtitleFontSize;
+            dialogueText.color = config.subtitleColor;
+            BackgroundImage.gameObject.SetActive(config.ShowBackground);
+            BackgroundImage.color = config.BackGroundColor;
+        }
+
+
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: Extra whitespace

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+namespace Studio23.SS2.DialogueSystem.Data
+{
+    public class DialogueBase : Node
+    {
+
+        [HideInInspector]
+        public string ID;
+
+        [Header("Public Properties")]
+        public string CharacterID;
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs:14]: We could have a per character scriptableobject. Less error prone than strings

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+        public string CharacterReaction;
+        [SerializeField]private CharacterTable.CharacterInfo _characterInfo;
+
+
+        [TextArea(3, 10)]
+        public string DialogueText;
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueBase.cs:20]: Will need to replace with localization system's serialized string.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        [SerializeField]private DialogueBase _startNode;
+        [SerializeField]private DialogueBase _currentNode;
+
+        public CharacterTable CharacterTable {  get; private set; }
+
+        public DialogueEvent OnDialogueStart;
+        public DialogueEvent OnDialogueNext;
+        public DialogueEvent OnDialogueComplete;
+
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/xNode/DialogueGraph.cs:15]: The API should contain an event when a dialogue option is encountered so that the dialogueUI class can show it. Additionally, there should be some way to tell the graph/node that an option was picked and that it should progress.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+            ApplyConfiguration();
+        }
+
+        private void ApplyConfiguration()
+        {
+            dialogueText.fontSize = config.subtitleFontSize;
+            dialogueText.color = config.subtitleColor;
+            BackgroundImage.gameObject.SetActive(config.ShowBackground);
+            BackgroundImage.color = config.BackGroundColor;
+        }
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueUI.cs]: UI-related changes and setups should be segregated to respective Ui controllers rather than having the functional script handle it.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/UITabs/DialogueGraphEditorUI.cs [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+    {
+        private string _csvFilePath;
+        private CharacterTable _characterTable;
+        private DialogueGraph _dialogueGraph;
+
+
+
+        public string GraphName;
+
+        private string folderPath = $"Assets/Resources/DialogueSystem/Graphs/";
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/UITabs/DialogueGraphEditorUI.cs:16]: naming convention for private var not followed.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/UITabs/DialogueGraphEditorUI.cs [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+            }
+            AssetDatabase.CreateAsset(_dialogueGraph, $"{folderPath}/{GraphName}/{GraphName}.asset");
+            AssetDatabase.SaveAssets();
+            AssetDatabase.Refresh();
+        }
+
+
+
+
+        private void PopulateDialogueGraph()
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/UITabs/DialogueGraphEditorUI.cs:106]: This populates from a csv file. It should be renamed to PopulateDialogueGraphFromCSV for clarity

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/Data/DialogueBaseNodeEditor.cs [hunk 27]
```diff
@@ -8,7 +8,7 @@ namespace Studio23.SS2.DialogueSystem.Data
     public class DialogueBaseNodeEditor : NodeEditor
     {
 
-        public Color TextColor= Color.white;
+        public Color TextColor = Color.white;
         private DialogueBase dialogueBase;
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/Data/DialogueBaseNodeEditor.cs:12]: Private var naming convention not followed

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueUI.cs [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+                DialogueText.text += text[i];
+                await UniTask.Delay((int)(_config.letterDelay * 1000));
+                if (_dialogueState == DialogueState.FastForwarded)
+                {
+                    DialogueText.text = text;
+                    break;
+                }
+
+            }
+            await UniTask.Delay((int)(text.Length * 100 * _config.nextSentenceDelayMultiplier));
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueUI.cs:89]: Multiplication by 1000/100 is hardcoded. To get a particular unit of Time, TimeSpan.FromSeconds and other variants can be used:

` await UniTask.Delay(TimeSpan.FromSeconds(10), ignoreTimeScale: false);`

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/DialogueSystemWizard.cs [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+        private void OnGUI()
+        {
+
+            GUILayout.Box(_header, GUILayout.Height(200), GUILayout.ExpandWidth(true));
+
+            _currentTab = GUILayout.Toolbar(_currentTab, new string[] { "Character Table", "Dialogue Graph" });
+
+            EditorGUILayout.BeginVertical(EditorStyles.helpBox);
+
+            if (_currentTab == 0)
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Editor/DialogueSystemWizard.cs:37]: Hardcoded tab index magic numbers. Should be refactored into variables.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueUI.cs [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+            OnDialoguePlayComplete?.Invoke();
+
+        }
+
+
+
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueUI.cs:95]: Unnecessary vertical space.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs [hunk 31]
```diff
@@ -24,11 +24,32 @@ public class CharacterReactionImage
...
+                        return false;
+                }
+
+                return true; // All checks passed, the CharacterTable is valid
+            }
+        }
+
+
+
         public List<CharacterInfo> characterList = new List<CharacterInfo>();
```
Comment (BDeshiDev) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/Models/CharacterTable.cs:48]: Public var naming rules not followed. Public access to this list is also unnecessary. If we want to later optimize it with a dictionary then the public list access would need to be removed.

**Review by BDeshiDev — APPROVED** [review]
The coupling  of the UI class with the manager is fixed. 

There are some minor changes such as incorrect var naming and  spacing issues. 

In a future version I  would recommend replacing the character table with an SO for each character info. It is possible to mess up typing CharacterID strings. With a scriptable, we can assign it and be sure that we have a correct reference without needing to write error checking code.
